content/doc/book/pipeline/pipeline-best-practices.adoc

```diff
@@ -0,0 +1,97 @@
+:hide-uri-scheme:
+:toc:
+endif::[]
+
+= Pipeline Best Practices
+
+== General
+
+=== Groovy Code in Pipelines as Glue
+
+Make sure to use Groovy code as a way of wiring together a set of actions rather than the main functionality. 
+This means that rather than relying on Pipeline functionality(groovy or Pipeline steps) to drive the build process forward, it is better to use single steps(such as `sh`) to accomplish multiple parts of the build. 
+As the complexity of Pipelines increase (in terms of amount of Groovy code, number of steps used, etc.), they require more resources (CPU, memory, storage) on the master.
+Think of Pipeline as a tool to accomplish a build rather than it being the core of that build.
+
+Example: Using a single maven build step to drive the build through its build/test/deploy process.
+
+=== Avoid complex Groovy code in Pipeline
+
+Groovy code within a Pipeline uses master resources(memory and CPU). 
+For a Pipeline, groovy code *always* executes on master which is why it does use master resources. 
+Because of this, the most critical action item is to reduce the number of groovy calls that need to be made (this includes the many imports of groovy methods being used). 
+The most common example methods are things such as:
+. *JsonSlurper:* This function (and other similar ones like XmlSlurper or ReadFile) would be to read in some file which is on disk into a JSON object within the Pipeline so it would look something like this: `def JsonFile = new JsonSlurper().parseText(readFile("$LOCAL_FILE")))`. The issue here is that this loads that file into memory 2x on the master and if that file is very large or this command is being done very often then it will drain a lot of unnecessary memory
+.. Solution: Instead of using JsonSlurper, use a shell step and return the standard out. This shell would look something like this: `def JsonReturn = sh label: '', returnStdout: true, script: 'echo "$LOCAL_FILE"| jq "."'`. This will use agent resources to read the file and even parse the file with a good `jq` filter
+. *HttpRequest:* Many times this would be used to grab data from an external source and store it into a variable. This is not a good idea to use because not only is that request coming directly from the master(which could give incorrect results for things like HTTPS requests if the master does not have certificates loaded) but also the return is again stored twice.
+.. Solution: Again using the shell step but this time with a `curl` command is the best direction to go.Then have the request come from master and also the data size can be shrunk with filtering.
+
+=== Reduce Repetition of similar Pipeline steps: 
+
+Combine Pipeline steps into single steps as often as possible to reduce the size of the flow-node graph(Pipeline’s way of tracking what step you are currently on). 
+This flow node graph has to be loaded onto the stack and if it is very large it can end up causing memory issues or even a StackOverflow.
```

Comment:
The only reason I mentioned the StackOverflow is because I saw it happen when trying to resume a pipeline on the master. But I do like your way better